### PR TITLE
Pulsar fix performance issue

### PR DESCRIPTION
**Pull request by rshermanTHG**
Fix to stop a new connection being created and closed for each message sent by camel. (Without my previous gitfu failure).

When testing this on my laptop utilising a slightly modified version of the example-spring-pulsar project it went from taking 89 seconds to send 10,000 messages with the current code to 7 seconds with this code. There was also a significant reduction in CPU usage on my laptop (i.e. the cpu fan didn't come on with the new code)

### CODE REVIEW

File: components/camel-pulsar/src/main/java/org/apache/camel/component/pulsar/PulsarProducer.java [hunk 1]
```diff
@@ -41,32 +42,40 @@ public PulsarProducer(PulsarEndpoint pulsarEndpoint) {
...
-        final ProducerBuilder<byte[]> producerBuilder = pulsarEndpoint
-            .getPulsarClient()
-            .newProducer()
-            .producerName(producerName)
-            .topic(topic);
-
-        if (properties != null && !properties.isEmpty()) {
-            producerBuilder.properties(properties);
+        if(producer == null) {
+            createProducer();
```
Comment (davsclaus) [inline components/camel-pulsar/src/main/java/org/apache/camel/component/pulsar/PulsarProducer.java]: Why not create the producer in the doStart method?

Comment (oscerd) [reply components/camel-pulsar/src/main/java/org/apache/camel/component/pulsar/PulsarProducer.java]: yes, probably it's better.

Comment (rshermanTHG) [reply components/camel-pulsar/src/main/java/org/apache/camel/component/pulsar/PulsarProducer.java]: I was following the pattern that is used by the camel-JMS component to lazily create the producer.

Comment (rshermanTHG) [reply components/camel-pulsar/src/main/java/org/apache/camel/component/pulsar/PulsarProducer.java]: I'll move it to doStart now

Comment (davsclaus) [reply components/camel-pulsar/src/main/java/org/apache/camel/component/pulsar/PulsarProducer.java]: Ah okay yeah so camel-jms is a bit more dynamic as its producers can be affinited to a request/reply pattern with temporary queues and all sorts of complexities

Comment (davsclaus) [reply components/camel-pulsar/src/main/java/org/apache/camel/component/pulsar/PulsarProducer.java]: If you want to offer that you can connect to pulsar on demand, then i suggest to add a new option on the endpoint which the end user can then turn on. And maybe have the option on the component level too, so you can set it generally

Comment (davsclaus) [reply components/camel-pulsar/src/main/java/org/apache/camel/component/pulsar/PulsarProducer.java]: As some ppl want to fail fast if connection issues, and others want it on-demand. And btw for pulsar does it have any kind of connection pooling concept with automatic re-connection or some sorts of that?

Comment (rshermanTHG) [reply components/camel-pulsar/src/main/java/org/apache/camel/component/pulsar/PulsarProducer.java]: When a pulsar client connects to pulsar there is built in support for multiple pulsar brokers and if one goes down automatically connecting to another. It does this using a zookeeper cluster that is used to keep track of all the meta data around pulsar. I understand that it gracefully handles connection issues.

Comment (rshermanTHG) [reply components/camel-pulsar/src/main/java/org/apache/camel/component/pulsar/PulsarProducer.java]: Have moved the creation of the Pulsar producer into the doStart method.

**Comment by oscerd** [comment]
Merged on master. Thanks

**Comment by davsclaus** [comment]
Isnt there a camel-pulsar on camel 2.x branch? And if so does it not have the same issue and we would need to backport the PR to it?

**Comment by oscerd** [comment]
Yes, @rshermanTHG will open a PR for that too, he said

**Comment by rshermanTHG** [comment]
> Isnt there a camel-pulsar on camel 2.x branch? And if so does it not have the same issue and we would need to backport the PR to it?

Yes and Yes. I'll get that done this afternoon.

**Pull request closed without merging**